planning/behavior_path_start_planner_module/src/start_planner_module.cpp

```diff
@@ -172,6 +202,41 @@ void StartPlannerModule::updateObjectsFilteringParams(
 
 void StartPlannerModule::updateData()
 {
+  // In PlannerManager::run(), it calls SceneModuleInterface::setData and
+  // SceneModuleInterface::setPreviousModuleOutput before module_ptr->run().
+  // Then module_ptr->run() invokes GoalPlannerModule::updateData and then
+  // planWaitingApproval()/plan(), so we can copy latest current_status to sp_planner_data_ here
+
+  // NOTE: onFreespacePlannerTimer copies sp_planner_data to its thread local variable, so we need
+  // to lock sp_planner_data_ here to avoid data race. But the following clone process is
+  // lightweight because most of the member variables of PlannerData/RouteHandler is
+  // shared_ptrs/bool
+  // begin of critical section
+  {
+    std::lock_guard<std::mutex> guard(sp_planner_data_mutex_);
+    if (!sp_planner_data_) {
```

Comment:
The if and else are deeply nested into the brackets, which makes the code a little harder to read when the `if (freespace_thread_status_)` line is reached (its an if inside and else , inside a lock_guard set of brackets)
I suggest using a lambda function []() -> StartPlannerData{}
So you can have an early return, something like this:
```
...
 sp_planner_data.value().update(
        *parameters_, *planner_data_, getCurrentStatus(), status_, isStopped());
return sp_planner_data;
}
auto & sp_planner_data = sp_planner_data_.value();
...
```

and you can eliminate the "else"

Reply:
Also, since the update method is called either way: 

sp_planner_data.update(
        *parameters_, *planner_data_, getCurrentStatus(), 
Do we really need to check if sp_planner_data_ has a value? 
cannot we always do this:
`sp_planner_data_ = StartPlannerData();
 sp_planner_data_.value().update(
        *parameters_, *planner_data_, getCurrentStatus(), status_, isStopped());
`
?
Then you only need to check for `if (freespace_thread_status_)` ?

Reply:
https://github.com/autowarefoundation/autoware.universe/pull/6779/commits/95ed1a51e694db2a8ff153cec78cf66325e22723

I moved the checking of `freespace_thread_status_` out of the `else` scope.

I do not want to refresh `start_planner_data_` value every time due to following reason:
1. PlannerData has the shared pointer to RouteHandler, which contains the shared pointer to `lanelet_map_ptr` and `routing_graph_ptr`. I'm making the deep copy of `start_planner_data.planner_data.route_handler` so StartPlanner can become the owner (== increase reference count) of `lanelet_map_ptr` and `routing_graph_ptr`.
2. But if we use huge lanelet map, this refresh of RouteHandler is heavy because it contains the vector of lanelets. So it is preferable to refresh it only when needed, or when the map/route has changed so to speak.
3. currently PlannerData does not have the flag indicating if `the map/route has changed`, but I will add it in the next PR. With that flag, we can skip the refresh of `StartPlannerData.planner_data.route_handler` if the flag is false.